## Practice Data Wrangling

![](https://iili.io/3S2kRyu.png)

## Pandas


Pandas is a Python library extensively used for data analysis and manipulation. It provides data structures like DataFrames, which are similar to tables, and Series, which are like lists or columns in a table. These structures allow for efficient handling and processing of structured data. Pandas is built on top of NumPy and integrates well with other scientific computing libraries. It simplifies tasks such as data cleaning, transformation, and analysis, making it a fundamental tool for data scientists and analysts.


### Installing and Importing Pandas

Installing Pandas
```
pip install pandas
```

In [1]:
!pip install pandas

importing pandas

In [2]:
## pd is widely used alias for pandas
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', 50)


### Creating Series and Pandas Data Frame

#### 1. Creating Series
 A Pandas Series is a one-dimensional labeled array capable of holding data of any type. It is similar to a column in a spreadsheet or a SQL table. The labels for the Series are collectively known as the index. A Series can be created from a variety of data types, including lists, dictionaries, and NumPy arrays. 


In [3]:
### Ex-1-Task-1

s = None 
a = [1, 2, 3, 4, 5, 6]
# Task: create a series using list 
### BEGIN SOLUTION 
# YOUR CODE HERE
s = pd.Series(data=a)
raise NotImplementedError() 
### END SOLUTION



NotImplementedError: 

In [ ]:
print(s)
print(f'The type of the variable s is: {type(s)}')

0    1
1    2
2    3
3    4
4    5
5    6
dtype: int64
The type of the variable s is: <class 'pandas.core.series.Series'>


In [ ]:
### Ex-1-Task-2
a = {
    'a': 80,
    'b': 90,
    'c': 100
}
s = None 
## Task: Create a series using dictionary

### BEGIN SOLUTION 
# YOUR CODE HERE
s = pd.Series(data=a, index=['a', 'b', 'c'])
raise NotImplementedError() 
### END SOLUTION


The type of the variable s is: <class 'pandas.core.series.Series'>


In [4]:
print(s)
print(f'The type of the variable s is: {type(s)}')

0    1
1    2
2    3
3    4
4    5
5    6
dtype: int64
The type of the variable s is: <class 'pandas.core.series.Series'>


In [ ]:
### Ex-1-Task-3
a = [1, 2, 3, 4, 5, 6]
a = np.array(a)
s = None
## Creating a series using numpy array
### BEGIN SOLUTION 
# YOUR CODE HERE
pd.Series(data = a)
raise NotImplementedError() 
### END SOLUTION


0    1
1    2
2    3
3    4
4    5
5    6
dtype: int64

In [6]:
print(f'The type of the variable s is: {type(s)}')


The type of the variable s is: <class 'NoneType'>


In [ ]:
### Ex-1-Task-4
population = {
    'Nepal': 30000000,
    'India': 1463865525,
    'Pakistan':255219554,
    'China': 1416096094
}
s = None
## Create a series with name 'population' and dtype='int64' 
### BEGIN SOLUTION 
# YOUR CODE HERE
s = pd.Series(population)

raise NotImplementedError() 
### END SOLUTION

In [57]:
print(s.head())
print(f'The type of the variable s is: {type(s)}')


Nepal         30000000
India       1463865525
Pakistan     255219554
China       1416096094
dtype: int64
The type of the variable s is: <class 'pandas.core.series.Series'>


#### 2. Creating DataFrame

In [ ]:
### Ex-2-Task-1
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
df = None 
# Creating Data Frame from Python List
### BEGIN SOLUTION 
# YOUR CODE HERE
df = pd.DataFrame(data=a, columns=['number'])
raise NotImplementedError() 
### END SOLUTION


In [62]:
print(df.head())
print(f'The type of the variable df is: {type(df)}')


   number
0       1
1       2
2       3
3       4
4       5
The type of the variable df is: <class 'pandas.core.frame.DataFrame'>


In [ ]:
### Ex-2-Task-2
my_list = [['A', 1], ['B', 2], ['C', 3], ['D', 4]]
# Creating Data Frame from List of List
### BEGIN SOLUTION 
# YOUR CODE HERE
df = pd.DataFrame(my_list, columns=['alphabet', 'number']).set_index('alphabet')
raise NotImplementedError() 
### END SOLUTION

In [68]:
print(df.head())
print(f'The type of the variable df is: {type(df)}')


          number
alphabet        
A              1
B              2
C              3
D              4
The type of the variable df is: <class 'pandas.core.frame.DataFrame'>


In [ ]:
### Ex-2-Task-3
my_data = {
    'country': ['Nepal', 'India', 'China'],
    'Population': [30000000, 1463865525, 1416096094]
}
df = None
# Creating Data Frame using Python Dictionary
### BEGIN SOLUTION 
# YOUR CODE HERE
df = pd.DataFrame(my_data, index=(['a', 'b', 'c']))
raise NotImplementedError() 
### END SOLUTION


In [76]:
print(df.head())
print(f'The type of the variable df is: {type(df)}')


  country  Population
a   Nepal    30000000
b   India  1463865525
c   China  1416096094
The type of the variable df is: <class 'pandas.core.frame.DataFrame'>


## Data Loading and Inspection

### Reading Data
A simple way to store big data is to use CSV files (comma Seperated files) or Parquet Files. 
- df_sales --> computed_insight_success_of_active_sellers.csv'
- df_product --> summer-products-with-rating-and-performance_2020-08.csv

In [ ]:
### Ex-3-Task-1
df_sales = None
df_product = None

# reading CSV file
### BEGIN SOLUTION 
# YOUR CODE HERE
df_sales = pd.read_csv("sales.csv")
df_product = pd.read_csv("products.csv")
raise NotImplementedError() 
### END SOLUTION


In [10]:
print(df_sales.head())
print(df_product.head())


                 merchantid  listedproducts  totalunitssold  \
0  5357bcf2bb72c5504882e889               2          120000   
1  5708773c3c02161b3f8c7900               5          107100   
2  5417aada4ad3ab27e954b76c               2          100007   
3  570f3a713a698c14278bb51e               1          100000   
4  53082ea15aefb07dfe1f2a4f               1          100000   

   meanunitssoldperproduct  rating  merchantratingscount  meanproductprices  \
0                  60000.0   4.219              320031.0               9.00   
1                  21420.0   3.934              139223.0               7.76   
2                  50004.0   4.053              108048.0               8.00   
3                 100000.0   3.889               19248.0               5.67   
4                 100000.0   4.036              366898.0               5.00   

   meanretailprices  averagediscount  meandiscount  meanproductratingscount  \
0              20.0             54.0          54.0                 

### Basic Inspection

In [ ]:
### Ex-3-Task-2
head_df = None 
tail_df = None 
desribe_num_df = None
desribe_cate_df = None
# check the first 10 records of the sales dataframe 
# check the last 10 records of the product dataframe
## Check the description of the numerical and categorical columns of prodcut dataframe
### BEGIN SOLUTION 
# YOUR CODE HERE
head_df = df_sales.head(10)
tail_df = df_product.tail(10)
df_product.describe()
raise NotImplementedError() 
### END SOLUTION

price  retail_price     units_sold  uses_ad_boosts       rating  \
count  1573.000000   1573.000000    1573.000000     1573.000000  1573.000000   
mean      8.325372     23.288620    4339.005086        0.432931     3.820896   
std       3.932030     30.357863    9356.539302        0.495639     0.515374   
min       1.000000      1.000000       1.000000        0.000000     1.000000   
25%       5.810000      7.000000     100.000000        0.000000     3.550000   
50%       8.000000     10.000000    1000.000000        0.000000     3.850000   
75%      11.000000     26.000000    5000.000000        1.000000     4.110000   
max      49.000000    252.000000  100000.000000        1.000000     5.000000   

       rating_count  rating_five_count  rating_four_count  rating_three_count  \
count   1573.000000        1528.000000        1528.000000         1528.000000   
mean     889.659250         442.263743         179.599476          134.549738   
std     1983.928834         980.203270         400.516231          311.690656   
min        0.000000           0.000000           0.000000            0.000000   
25%       24.000000          12.000000           5.000000            4.000000   
50%      150.000000          79.000000          31.500000           24.000000   
75%      855.000000         413.500000         168.250000          129.250000   
max    20744.000000       11548.000000        4152.000000         3658.000000   

       rating_two_count  rating_one_count  badges_count  badge_local_product  \
count       1528.000000       1528.000000   1573.000000          1573.000000   
mean          63.711387         95.735602      0.105531             0.018436   
std          151.343933        214.075544      0.340709             0.134565   
min            0.000000          0.000000      0.000000             0.000000   
25%            2.000000          4.000000      0.000000             0.000000   
50%           11.000000         20.000000      0.000000             0.000000   
75%           62.000000         94.000000      0.000000             0.000000   
max         2003.000000       2789.000000      3.000000             1.000000   

       badge_product_quality  badge_fast_shipping  \
count            1573.000000          1573.000000   
mean                0.074380             0.012715   
std                 0.262472             0.112075   
min                 0.000000             0.000000   
25%                 0.000000             0.000000   
50%                 0.000000             0.000000   
75%                 0.000000             0.000000   
max                 1.000000             1.000000   

       product_variation_inventory  shipping_option_price  \
count                  1573.000000            1573.000000   
mean                     33.081373               2.345200   
std                      21.353137               1.024371   
min                       1.000000               1.000000   
25%                       6.000000               2.000000   
50%                      50.000000               2.000000   
75%                      50.000000               3.000000   
max                      50.000000              12.000000   

       shipping_is_express  countries_shipped_to  inventory_total  \
count          1573.000000           1573.000000      1573.000000   
mean              0.002543             40.456453        49.821360   
std               0.050379             20.301203         2.562799   
min               0.000000              6.000000         1.000000   
25%               0.000000             31.000000        50.000000   
50%               0.000000             40.000000        50.000000   
75%               0.000000             43.000000        50.000000   
max               1.000000            140.000000        50.000000   

       has_urgency_banner  merchant_rating_count  merchant_rating  \
count               473.0           1.573000e+03      1573.000000   
mean                  1.0           2.649583e+04     

In [12]:
df_sales.head()

merchantid  listedproducts  totalunitssold  \
0  5357bcf2bb72c5504882e889               2          120000   
1  5708773c3c02161b3f8c7900               5          107100   
2  5417aada4ad3ab27e954b76c               2          100007   
3  570f3a713a698c14278bb51e               1          100000   
4  53082ea15aefb07dfe1f2a4f               1          100000   

   meanunitssoldperproduct  rating  merchantratingscount  meanproductprices  \
0                  60000.0   4.219              320031.0               9.00   
1                  21420.0   3.934              139223.0               7.76   
2                  50004.0   4.053              108048.0               8.00   
3                 100000.0   3.889               19248.0               5.67   
4                 100000.0   4.036              366898.0               5.00   

   meanretailprices  averagediscount  meandiscount  meanproductratingscount  \
0              20.0             54.0          54.0                   8836.0   
1              34.2             61.0          61.0                   4010.0   
2               8.0             -1.0          -1.0                   5531.0   
3              19.0             71.0          71.0                  18393.0   
4              33.0             85.0          85.0                  13789.0   

   totalurgencycount  urgencytextrate  
0                1.0             50.0  
1                3.0             60.0  
2                NaN              NaN  
3                1.0            100.0  
4                NaN              NaN

In [ ]:
### Ex-3-Task-3

# set merchant id as the index of the sales dataframe and check index
### BEGIN SOLUTION 
# YOUR CODE HERE
df_sales.set_index('merchantid')
raise NotImplementedError() 
### END SOLUTION

listedproducts  totalunitssold  \
merchantid                                                 
5357bcf2bb72c5504882e889               2          120000   
5708773c3c02161b3f8c7900               5          107100   
5417aada4ad3ab27e954b76c               2          100007   
570f3a713a698c14278bb51e               1          100000   
53082ea15aefb07dfe1f2a4f               1          100000   
...                                  ...             ...   
567bad8582c35f28103e5813               1               2   
5b24edf0d1c360301099e9c6               1               2   
5d417e7070327a2743021677               1               1   
5add589edb5f1f32114d07df               1               1   
5d4b8931933fb13f8a1fe232               1               1   

                          meanunitssoldperproduct  rating  \
merchantid                                                  
5357bcf2bb72c5504882e889                  60000.0   4.219   
5708773c3c02161b3f8c7900                  21420.0   3.934   
5417aada4ad3ab27e954b76c                  50004.0   4.053   
570f3a713a698c14278bb51e                 100000.0   3.889   
53082ea15aefb07dfe1f2a4f                 100000.0   4.036   
...                                           ...     ...   
567bad8582c35f28103e5813                      2.0   3.999   
5b24edf0d1c360301099e9c6                      2.0   4.303   
5d417e7070327a2743021677                      1.0   3.598   
5add589edb5f1f32114d07df                      1.0   3.956   
5d4b8931933fb13f8a1fe232                      1.0   4.007   

                          merchantratingscount  meanproductprices  \
merchantid                                                          
5357bcf2bb72c5504882e889              320031.0               9.00   
5708773c3c02161b3f8c7900              139223.0               7.76   
5417aada4ad3ab27e954b76c              108048.0               8.00   
570f3a713a698c14278bb51e               19248.0               5.67   
53082ea15aefb07dfe1f2a4f              366898.0               5.00   
...                                        ...                ...   
567bad8582c35f28103e5813               21307.0              11.00   
5b24edf0d1c360301099e9c6                 693.0              16.00   
5d417e7070327a2743021677                 378.0               2.72   
5add589edb5f1f32114d07df                1134.0              14.00   
5d4b8931933fb13f8a1fe232                1380.0              11.00   

                          meanretailprices  averagediscount  meandiscount  \
merchantid                                                                  
5357bcf2bb72c5504882e889              20.0             54.0          54.0   
5708773c3c02161b3f8c7900              34.2             61.0          61.0   
5417aada4ad3ab27e954b76c               8.0             -1.0          -1.0   
570f3a713a698c14278bb51e              19.0             71.0          71.0   
53082ea15aefb07dfe1f2a4f              33.0             85.0          85.0   
...                                    ...              ...           ...   
567bad8582c35f28103e5813              10.0            -10.0         -10.0   
5b24edf0d1c360301099e9c6              16.0              0.0           0.0   
5d417e7070327a2743021677               9.0             70.0          70.0   
5add589edb5f1f32114d07df              14.0              0.0           0.0   
5d4b8931933fb13f8a1fe232              30.0             64.0          64.0   

                          meanproductratingscount  totalurgencycount  \
merchantid                                                             
5357bcf2bb72c5504882e889                   8836.0                1.0   
5708773c3c02161b3f8c7900                   4010.0                3.0   
5417aada4ad3ab27e954b76c                   5531.0                NaN   
570f3a713a698c14278bb51e                  18393.0                1.0   
53082ea15aefb07dfe1f2a4f                  13789.0                NaN   
...                                           .

In [14]:
print(f'Index: {df_sales.index}')

Index: RangeIndex(start=0, stop=958, step=1)


In [ ]:
### Ex-3-Task-4
sorted_df = None
## Sort the sales dataframe by rating in Ascending Order, if two records have same rating then sort in descending order by total unit sold
### BEGIN SOLUTION 
# YOUR CODE HERE
df_sales.sort_values(
    by = ['rating', 'totalunitssold'],
    ascending=[True, False]
)
raise NotImplementedError() 
### END SOLUTION

merchantid  listedproducts  totalunitssold  \
927  5e63469b2fdc774466e15dd5               1              10   
704  5e8d923d20c9b7545017f1f9               1             100   
705  5e6cb9806bab2075b2b64d15               1             100   
706  5e8c98966642021100f8db13               1             100   
707  5e537a4038654d5bb8db71b4               1             100   
..                        ...             ...             ...   
948  574d559cb06e205d099d9107               1              10   
949  5db248327c0d0800c24fc084               1              10   
887  5b07dc937752c81ecb291b66               1             100   
888  559f372b577ce1196ef7f66c               1             100   
889  5dab9ece69ce753090f6735c               1             100   

     meanunitssoldperproduct  rating  merchantratingscount  meanproductprices  \
927                     10.0   2.333                   3.0               5.75   
704                    100.0   2.941                  17.0              11.00   
705                    100.0   3.000                   3.0              12.00   
706                    100.0   3.034                  29.0              12.00   
707                    100.0   3.039                 231.0               5.00   
..                       ...     ...                   ...                ...   
948                     10.0   4.501                2038.0              18.00   
949                     10.0   4.513                 160.0               5.68   
887                    100.0   4.522                 343.0               3.73   
888                    100.0   4.578                1032.0               8.00   
889                    100.0   5.000                   0.0               7.00   

     meanretailprices  averagediscount  meandiscount  meanproductratingscount  \
927               5.0            -15.0         -15.0                      0.0   
704              50.0             78.0          78.0                      9.0   
705              11.0             -9.0          -9.0                      1.0   
706              75.0             84.0          84.0                     28.0   
707               5.0              0.0           0.0                      6.0   
..                ...              ...           ...                      ...   
948              59.0             70.0          70.0                      6.0   
949               5.0            -13.0         -13.0                      2.0   
887              17.0             79.0          79.0                     27.0   
888              47.0             83.0          83.0                    110.0   
889               6.0            -16.0         -16.0                      0.0   

     totalurgencycount  urgencytextrate  
927                NaN              NaN  
704                NaN              NaN  
705                NaN              NaN  
706                NaN              NaN  
707                NaN              NaN  
..                 ...              ...  
948                NaN              NaN  
949                NaN              NaN  
887                NaN              NaN  
888                NaN              NaN  
889                NaN              NaN  

[958 rows x 13 columns]

In [ ]:
### Ex-3-Task-5
portion_df = None
### Get the sorted dataframe for which rating is 4.041 and store in portion_df
### BEGIN SOLUTION 
portion_df = df_sales[df_sales['rating'] == 4.041]
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

In [27]:
portion_df.head()

merchantid  listedproducts  totalunitssold  \
307  5810b1db3deb94160f5cd4a3               1            5000   
308  58380161e18c0f0c70dd6907               1            5000   
518  53e98c6446188e1d4d3a38bd               1            1000   
631  580c7d0a8b67fa197454a2d1               3             300   
668  5eba4bab58ed6ba265685ba1               2             200   

     meanunitssoldperproduct  rating  merchantratingscount  meanproductprices  \
307                   5000.0   4.041               15436.0                6.0   
308                   5000.0   4.041               14194.0                8.0   
518                   1000.0   4.041                4230.0                1.8   
631                    100.0   4.041                4705.0                6.0   
668                    100.0   4.041                 123.0                8.0   

     meanretailprices  averagediscount  meandiscount  meanproductratingscount  \
307               6.0              0.0           0.0                    616.0   
308               7.0            -14.0         -14.0                   1248.0   
518               2.0             10.0          10.0                     76.0   
631               6.0              0.0           0.0                     27.0   
668              10.0             20.0          20.0                     58.0   

     totalurgencycount  urgencytextrate  
307                NaN              NaN  
308                NaN              NaN  
518                NaN              NaN  
631                1.0             33.0  
668                1.0             50.0

### Data Types and Conversion

In [29]:
### Ex-4-Task-1

# the price of the product is upto 2 digit place. I want to convert it into int
### BEGIN SOLUTION 
# YOUR CODE HERE
df_product['price']=df_product['price'].astype('int')
# raise NotImplementedError() 
### END SOLUTION

In [30]:
df_product['price']

0       16
1        8
2        8
3        8
4        2
        ..
1568     6
1569     2
1570     5
1571    13
1572     7
Name: price, Length: 1573, dtype: int64

In [ ]:
### Ex-4-Task-2
# The datatype of the craw_month is Object type. I want to convert it to data time format and store it in same columns
### BEGIN SOLUTION 
# YOUR CODE HERE
df_product['crawl_month'] = pd.to_datetime(df_product['crawl_month'])
raise NotImplementedError() 
### END SOLUTION

In [33]:
df_product['crawl_month']

0      2020-08-01
1      2020-08-01
2      2020-08-01
3      2020-08-01
4      2020-08-01
          ...    
1568   2020-08-01
1569   2020-08-01
1570   2020-08-01
1571   2020-08-01
1572   2020-08-01
Name: crawl_month, Length: 1573, dtype: datetime64[ns]

## Data Selection and Filtering

### Selecting Columns and Rows

In [ ]:
### Ex-5-Task-1
ser = None 

# Select a 'title' column from product dataframe and store in separate variables as series (ser)
### BEGIN SOLUTION 
# YOUR CODE HERE
ser = df_product['title']
raise NotImplementedError() 
### END SOLUTION

In [35]:
type(ser)

pandas.core.series.Series

In [ ]:
### Ex-5-Task-2
product_price = None 
# Select multiple columns ('title', 'price', and 'retail price' ) from product dataframe and store in variable product_price.
### BEGIN SOLUTION 
# YOUR CODE HERE
product_price = df_product[['title', 'price', 'retail_price']]
raise NotImplementedError() 
### END SOLUTION

In [39]:
type(product_price)

pandas.core.frame.DataFrame

### Selection by labels and Selection by position

In [ ]:
### Ex-5-Task-3
single_data = None
## Select a data of title column and at index 1572 of product dataframe
### BEGIN SOLUTION 
# YOUR CODE HERE
single_data = df_product['title'][1572]
raise NotImplementedError() 
### END SOLUTION


In [43]:
df_product['title']

0       2020 Summer Vintage Flamingo Print  Pajamas Se...
1       SSHOUSE Summer Casual Sleeveless Soirée Party ...
2       2020 Nouvelle Arrivée Femmes Printemps et Été ...
3       Hot Summer Cool T-shirt pour les femmes Mode T...
4       Femmes Shorts d'été à lacets taille élastique ...
                              ...                        
1568    Nouvelle Mode Femmes Bohême Pissenlit Imprimer...
1569    10 couleurs femmes shorts d'été lacent ceintur...
1570    Nouveautés Hommes Siwmwear Beach-Shorts Hommes...
1571    Mode femmes d'été sans manches robes col en V ...
1572    Pantalon de yoga pour femmes à la mode Slim Fi...
Name: title, Length: 1573, dtype: object

In [41]:
single_data

'Pantalon de yoga pour femmes à la mode Slim Fit Fitness Running Leggings'

In [ ]:
### Ex-5-Task-4
single_data = None
## Select a data of 6th column and at  5tj row of product dataframe
### BEGIN SOLUTION 
single_data = df_product.iloc[4, 5]
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

In [54]:
print(single_data)

100


In [ ]:
### Ex-5-Task-5
sliced_data = None 
## Select a portion of data frame with column name currency_buyer, units_sold, merchant_rating_count, merchant_rating and with index 100 to 200
### BEGIN SOLUTION 
# YOUR CODE HERE
sliced_data = df_product.loc[100:200, ['currency_buyer', 'units_sold', 'merchant_rating_count','merchant_rating' ]]
raise NotImplementedError() 
### END SOLUTION

In [57]:
sliced_data.head()

currency_buyer  units_sold  merchant_rating_count  merchant_rating
100            EUR         100                    288         3.729167
101            EUR       10000                  54504         3.801684
102            EUR         100                     67         3.298507
103            EUR        5000                  99283         4.285598
104            EUR         100                    239         3.882845

In [ ]:
### Ex-5-Task-6
sliced_data = None 
## Select a portion of data frame with 3rd to 8th columns and with rows 100 to 200 
# it must only include 100 rows and 7 columns 
### BEGIN SOLUTION 
sliced_data = df_product.iloc[100:200, 3:8]
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION


In [60]:
sliced_data.head()

retail_price currency_buyer  units_sold  uses_ad_boosts  rating
100             8            EUR         100               1    3.18
101             8            EUR       10000               0    3.96
102            17            EUR         100               1    2.80
103             5            EUR        5000               0    3.89
104            65            EUR         100               0    3.97

### Boolean Indexing and Conditional Filtering

In [ ]:
### Ex-5-Task-7
mean = None
Q3 = None 
records = None 
# Find mean and 3rd quantile of the price from product dataframe and get all the records that lies between mean and Q3
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

In [ ]:
### Ex-5-Task-8
summer_product = None 
## Find all the row in product dataframe containing 'summer' keyword in its title. Make sure the string is lower to get all the records. 
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

## Handling Missing Data

In [ ]:
### Ex-6-Task-1
count_na = None
# Count the number of null data in each columns
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

In [ ]:
### Ex-6-Task-2

# Impute the  product color with maximum occur value as it is the categorical columns 
# Impute the rating five count by mean and rating four count by median of respective columns
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

## Data Cleaning

### Removing Duplicates

In [ ]:
### Ex-7-Task-1
duplicate_count = None
df_drop_duplicate = None # store the data frame after removing duplicates 
# Count no. of duplicates value in product dataframe and drop all the duplicate values 
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

## Data Transformation

### Applying Functions: Apply, map

In [ ]:
### Ex-7-Task-2

df_sales['total_profit'] = None
# Calculate the total profit using apply function in pandas and save in total_profit column
# Formulae: total_profit = (mean product price - mean retail price) * total units sold
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

### Binning and Categorizing Data

In [ ]:
### Ex-7-Task-3
df_product['sales_category'] = None
# Divide the row into three category based on their units sold and save it into the sales_category column of produdct dataframe
# category are : low_sales, medium sales, and high sales 
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

## Data Aggregation and Grouping

### Groupby aggregation function


In [ ]:
### Ex-8-Task-1

group_df = None 
# Create a group of low, medium and high sales and then calulcate, mean of price, median of retail price, min of units_sold and max of rating_five_count of each category
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

In [ ]:
### Ex-8-Task-2
group_df = None
## Calculate the Coefficient of Variation of each sales category
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION


In [ ]:
### Ex-8-Task-3
df_product['price_zscore'] = None
## Calculate the Z_score of price for each sales category
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

## Merging

In [ ]:
### Ex-9-Task-1
merged_df = None
# Merge the sales and product dataframe and delete all the non common entries. Merge two dataframe based on merchant id.
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

# Data Wrangling End to End 

## Data Cleaning

#### Features and Columns 
The data was scraped in the french localisation (hence some non-ascii latin characters such as « é » and « à ») in the title column.

The title_orig on the other hand contains the original title (the base title) that is displayed by default. When a translation is provided by the seller, it appears in the title column. When the title and title_orig columns are the same, it generally means that the seller did not specify a translation that would be displayed to users with french settings.

A picture is worth a thousand words. In the following screenshot you see some features and how to interpret them.

![](wish_viz.jpg)

In [ ]:
### Ex-10-Task-1
df = None 
# Read the product data frame with rating and performance 
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

There are total 43 columns. Let's try to reduce the number. 

#### Some considerations:
- Excluded product color and size because it's just showing the ones that were found in getting the data : there are more colors and sizes for each product so having just one for each category doesn't say much.
- Excluded currency used, as every price is in EUR
- Excluded shipping option, the "shipping_is_express" column is enough.
- Excluded product's and merchant's id's and pictures, that won't be needed.
- Use only original title 



After removal of column, it must only have these columns. 

'title_orig', 'price', 'retail_price',
'units_sold', 'uses_ad_boosts', 'rating', 'rating_count',
'rating_five_count', 'rating_four_count', 'rating_three_count',
'rating_two_count', 'rating_one_count', 'badges_count',
'badge_local_product', 'badge_product_quality', 'badge_fast_shipping',
'tags', 'shipping_option_price', 'shipping_is_express', 'countries_shipped_to',
'inventory_total', 'has_urgency_banner',
'merchant_rating_count', 'merchant_rating',
'merchant_has_profile_picture'

In [ ]:
### Ex-10-Task-2
df = df[['title_orig', 'price', 'retail_price',
        'units_sold', 'uses_ad_boosts', 'rating', 'rating_count',
       'rating_five_count', 'rating_four_count', 'rating_three_count',
       'rating_two_count', 'rating_one_count', 'badges_count',
       'badge_local_product', 'badge_product_quality', 'badge_fast_shipping',
       'tags', 'shipping_option_price','shipping_is_express', 'countries_shipped_to',
        'has_urgency_banner','merchant_rating_count', 'merchant_rating','merchant_has_profile_picture',
        ]]

# print(df.info())
# print(df.isna().sum())
# print(df['has_urgency_banner'].value_counts())
# Check the data frame information, total no. of null entries in each columns and has_urgency_banner columns counts
# After loading the data we se we have 1573 entries and some null values, let's see what some of them are about:
# - Exactly 45 for rating counts from 1 to 5 star, could it be that there are products with no ratings at all?
# - 1100 null values in the has urgency banner : the values there are 1's so the null values should be turned into 0's
# Now, Rename the columns 'has_urgency_banner' to 'is_running_out' and 'title_orig' to 'title' to make it more clear.
# **Fix the runing out column. fill all the null values with 0's.
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

#### Check if all the 45 null values for all rating columns belong to same product.
- Get the dataframe for which rating_five_count have null values
- Then, Count the no. of null values for four, three, two and one rating column.

Conclusion: As expected we see the null values in the rating are actually of the same 45 products

In [ ]:
### Ex-10-Task-3
df[df['rating_five_count'].isna()][['rating_four_count', 'rating_three_count',
       'rating_two_count', 'rating_one_count']].isna().sum()
df[df['rating_five_count'].isna()].head()
# They seem to be having a rating of 5 even with no ratings: we are going to change the rating and all the rating counts to 0
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

## Data Exploration

```df['units_sold'].unique()```

It seems like they are all numbers approximated to their lower bound. So let's pretend they are divided in ranges like this:

- 10-100
- 100-1000
- 1000-5000
- 5000-10000
- 10000-20000
- 20000-50000
- 50000-100000
- 100000+

Let's define 3 performance characteristics:

- Average : 0-1000
- Successful: 1000-20000
- Very Successful 20000+

Find the percentage of product belongs to each category.
Steps:
1. First find the performance category and store in performance columns
2. Then, Calculate the percentage of product in each performance category



In [ ]:
### Ex-10-Task-4
percentage_of_product_in_each_performance_category = None
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION


## Product Rating and Performance Comparision

Looking at the rating columns I notice two things:

- Some products have a rating count that's too low to be considered, let's only look at products with more than 150 ratings to try to have a fair result
- The rating of the products is a decimal number: let's create another column that categorizes each product into it's rating range.
    - columns are 
        1. < 1 if rating < 1
        2. 1-2 if rating >= 1 and rating < 2
        3. 2-3 if rating >= 2 and rating < 3
        4. 3-4 if rating >= 3 and rating < 4
        5. 4-5 if rating >= 4 and rating < 5
- Create pivot table with performance in index, raiting_range in column and no. of product as values



In [ ]:
### Ex-10-Task-5
df_ratings = pd.DataFrame()  ## get records with rating_count > 100
df_ratings['rating_range'] = None
pivot_table = None
### BEGIN SOLUTION 
# YOUR CODE HERE
raise NotImplementedError() 
### END SOLUTION

This table shows that if the product have average rating of 3-4. THen, it is most likely to be successful product.